### PR TITLE
Implement parsing of the Amount from the Transaction ID 

### PR DESCRIPTION
**Pull request by yyosifov**
* For EUR transactions - try parsing the amount from the Transaction Id 
* Added unit tests for the regex and the prepare transaction functions

Closes #1462 (from the frontend repo)

### CODE REVIEW

**Review by igoychev — CHANGES_REQUESTED** [review]
Amazing job with the tests and the corner cases!!!
Please see my two comments regarding !=BGN and additional safety with marking the transaction as unrecognized if parsing didn't work.

**Comment by sashko9807** [comment]
Assuming that the position of the amount is the exact same on every transactionId, and the amount is always represented as a decimal number, can't we just use `transactionId.slice(-16, -9)`.  Should be faster than doing regex matching on the string.

File: apps/api/src/tasks/bank-import/import-transactions.task.ts [hunk 1]
```diff
@@ -298,8 +307,23 @@ export class IrisTasks {
...
         .match(this.regexPaymentRef)
 
+      const transactionAmount = {
+        amount: trx.transactionAmount?.amount,
+        currency: trx.transactionAmount?.currency,
+      }
+      const id = trx.transactionId?.trim() || ''
+
+      // If we receive EUR in the BGN account - try parsing from the transaction id the amount in BGN
+      if (trx.transactionAmount?.currency === Currency.EUR && trx.transactionAmount?.amount > 0) {
```
[the remark this answers is not in the document]
Comment (yyosifov) [reply apps/api/src/tasks/bank-import/import-transactions.task.ts:317]: @igoychev  Sure, I can do that. I'm worried that we don't know if the same parsing rule will apply + I guess we haven't seen any != EUR/BGN transactions, so we might want to actually see these and then handle them correctly? I'm worried that this might hide some issues and automatically "resolve" them. Let me know what you think

File: apps/api/src/tasks/bank-import/import-transactions.task.ts [hunk 2]
```diff
@@ -298,8 +307,23 @@ export class IrisTasks {
...
+        const amount = this.extractAmountFromTransactionId(id, trx.valueDate)
+        if (amount) {
+          transactionAmount.amount = amount
+          transactionAmount.currency = Currency.BGN
+        }
```
[the remark this answers is not in the document]
Comment (yyosifov) [reply apps/api/src/tasks/bank-import/import-transactions.task.ts:322]: Currently the "logic" is - if we can parse the Amount and it's a positive one (> 0) then we set update the transaction. Otherwise, we won't touch it and it will be recorded as received - therefore not transformed and we can see what's the case and how to handle it. It's kind of defensive - if the code is not sure, it doesn't change anything.

Again, that's just my concern here.

File: apps/api/src/tasks/bank-import/import-transactions.task.ts [hunk 3]
```diff
@@ -309,8 +333,8 @@ export class IrisTasks {
         senderIban: trx.debtorAccount?.iban?.trim(),
         recipientIban: trx.creditorAccount?.iban?.trim(),
         type: trx.creditDebitIndicator === 'CREDIT' ? 'credit' : 'debit',
-        amount: toMoney(trx.transactionAmount?.amount),
-        currency: trx.transactionAmount?.currency,
+        amount: toMoney(transactionAmount.amount),
+        currency: transactionAmount.currency,
         description: trx.remittanceInformationUnstructured?.trim(),
         // Not saved in the DB, it's added only for convinience and efficiency
         matchedRef: matchedRef ? matchedRef[0] : null,
```
[the remark this answers is not in the document]
Comment (yyosifov) [reply apps/api/src/tasks/bank-import/import-transactions.task.ts:340]: @igoychev  If we don't parse the transaction amount (i.e. it's a valid number > 0) I've left it to not change the transaction amount and currency at all. I thought this was a good way to protect us from handling cases that we're not sure how to handle. My concern is that it might be a valid transaction, but the amount is not in the id where expect it.

Let me know what you think?

[the remark this answers is not in the document]
Comment (igoychev) [reply apps/api/src/tasks/bank-import/import-transactions.task.ts:340]: > If we don't parse the transaction amount (i.e. it's a valid number > 0) I've left it to not change the transaction amount and currency at all. I thought this was a good way to protect us from handling cases that we're not sure how to handle.

Yes, this is ok, just for this case, we also need to set the matchedRef to null so that the record is considered as unrecognized

File: apps/api/src/tasks/bank-import/import-transactions.task.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (igoychev) [reply apps/api/src/tasks/bank-import/import-transactions.task.ts:317]: Well, because the receiving account is BGN then it is safe to assume, that the amount coming in the transactionId is always BGN for all types of incoming currencies.

File: apps/api/src/tasks/bank-import/import-transactions.task.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (yyosifov) [reply apps/api/src/tasks/bank-import/import-transactions.task.ts:340]: While writing a test I hit a corner case here. If it's EUR + not recognized, as per a previous comment we always set "BGN" as Currency and the amount would be the only one we have - the EUR amount, and it will be confusing.
So I'd suggest that we leave the 'BGN' currency override only when we can parse it and leave the original currency + mark as unrecognized in the other case?

[the remark this answers is not in the document]
Comment (igoychev) [reply apps/api/src/tasks/bank-import/import-transactions.task.ts:340]: yes, that's the idea - leave everything as it came

[the remark this answers is not in the document]
Comment (yyosifov) [reply apps/api/src/tasks/bank-import/import-transactions.task.ts:343]: Ok it's updated now.

File: apps/api/src/tasks/bank-import/import-transactions.task.spec.ts [hunk 4]
```diff
@@ -544,6 +545,112 @@ describe('ImportTransactionsTask', () => {
...
+          'Booked_6516347588_70001524349032963FTRO23184809601C2023010361.12_20230103',
+        bookingDate: '2023-01-03',
+        creditorAccount: {
+          iban: 'BG66UNCR70001524349032',
+        },
+        creditorName: 'СДРУЖЕНИЕ ПОДКРЕПИ БГ',
+        debtorAccount: {
+          iban: 'BG21UNCR111111111111',
+        },
+        debtorName: 'Name not relevant for the example',
```
Comment (igoychev) [inline apps/api/src/tasks/bank-import/import-transactions.task.spec.ts:560]: Wow! Great example for setting the high bar of how test data should look like!

**Review by igoychev — APPROVED** [review]
Great and special thanks again for the extra safety!

**Comment by yyosifov** [comment]
@sashko9807 Sorry, I just saw your comment. Here are my thoughts:

- Regarding performance - I doubt it will be a big difference for so short string and the operation is quite rare - i.e. only happens on a new transaction. I would agree that we should improve readability/maintainability if there is a way :)
- The slice you suggested looks cool, but it will work only if we have 7 symbols in the Amount. For example, if someone donated 100 EUR it won't parse it correctly. Example: `'Booked_6516347588_70001524349032963FTRO23184809601C20230703100.00_20230703'.slice(-16, -9)' = '3100.00'` which is incorrect

**Review by igoychev — APPROVED** [review]
